### PR TITLE
Add self-signed option to allow configure with private gitlab instances

### LINKED ISSUE #27: Configure fails when running against an instance with a self-signed certificate

### PR DESCRIPTION
**Pull request by jlengrand**
* Also add support for command options

Should be merged after #31 (and the Pr should be updated to be against master).

Fixes #27

### CODE REVIEW

**Comment by jlengrand** [comment]
Some commands from @RamonGebben that I still have to process : 

* The option should be at software level, not command level (nothing should work without the flag)
* Use index for command level options instead of weird command indexing.

I will leave the command level option logic so we can reuse it later.

File: index.js [hunk 1]
```diff
@@ -52,11 +59,21 @@ const run = async() => {
...
   }
 
-  commands.forEach(({ trigger, description, fn }) => {
+  commands.forEach(({ trigger, description, fn, options }) => {
     program
       .command(trigger)
       .description(description)
       .action((...args) => fn(config, ...args));
+
+    if(options && options.length > 0){
```
[the remark this answers is not in the document]
Comment (jlengrand) [reply index.js]: I ended up removing the logic since it is irrelevant at the moment

**Comment by jlengrand** [comment]
Alright, all feedback processed. @RamonGebben , this is coming back to you :).

File: lib/options/disableCertificateVerification/index.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const { logger } = require('../../utils/logger');
+const chalk = require('chalk');
+
+const disableCertificateVerification = async() => {
+    logger.log(chalk.red.bold('⚠️ Disabling certificate verification. This is unsafe and should only be used as last resort.'));
```
[the remark this answers is not in the document]
Comment (jlengrand) [reply lib/options/disableCertificateVerification/index.js]: good idea

File: lib/options/disableCertificateVerification/index.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+    process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
```
[the remark this answers is not in the document]
Comment (jlengrand) [reply lib/options/disableCertificateVerification/index.js]: I do kinda agree with the comment. Thinking about #11 for example, some of the options should definitely be part of the config. 

I think that this first version of the implementation is probably good enough for now though. Let's see what happens

File: lib/options/disableCertificateVerification/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jlengrand) [reply lib/options/disableCertificateVerification/index.js]: Alright, I gave it a go. Let me see if you are happy about the implementation :).

**Comment by RamonGebben** [comment]
Now the test coverage has dropped by 4.6% can you see if you can up that again as well?

File: lib/utils/logger/index.js [hunk 4]
```diff
@@ -1,3 +1,19 @@
...
+}
+
+Logger.prototype.err = function(...args) {
+  console.log(chalk.red.bold(...args));
+}
+
 module.exports = {
-  logger: console
-};
+  logger: new Logger()
```
Comment (RamonGebben) [inline lib/utils/logger/index.js]: I'm not sure if we need a constructor for this. I think we should just use an object literal since we don't need to access anything in the `this` context. 

So maybe like
 ```const logger = { warn: (...args) => console.log(...args) }```

That way we don't need to `new` it.

File: lib/commands/configure/index.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ const chalk = require('chalk');
 
 const configure = async() => {
   if(await checkConfigExists()){
-    logger.log(chalk.red.bold('⚠️ Mergify is already configured. Configuring again will override the existing file.'));
+    logger.warn('⚠️ Mergify is already configured. Configuring again will override the existing file.');
```
Comment (RamonGebben) [inline lib/commands/configure/index.js]: Since all warnings have a ⚠️ in front maybe also put those in the logger. 
Same goes for the Uh-oh cat 🙀.

File: lib/utils/logger/index.js [hunk 6]
```diff
@@ -1,3 +1,19 @@
+const chalk = require('chalk');
+
+var Logger = function(){};
```
Comment (RamonGebben) [inline lib/utils/logger/index.js]: Let's turn this into a getLogger function and pass a `transport` as the first argument. 
Which returns the object literal which I was talking about down below.

That way you can provide a mocked version of `console` in your tests. And naturally this would have more options for the future as well.

File: lib/commands/configure/index.js (hunk 5 above)
Comment (jlengrand) [reply lib/commands/configure/index.js]: Fixed

File: lib/utils/logger/index.js (hunk 6 above)
Comment (jlengrand) [reply lib/utils/logger/index.js]: No idea what you're talking about. This is all Chinese at the moment

File: lib/utils/logger/index.js (hunk 4 above)
Comment (jlengrand) [reply lib/utils/logger/index.js]: Hum, that's how I had it first but I didn't get it working. 
Let me try again

Comment (jlengrand) [reply lib/utils/logger/index.js]: That's what I get if I try to use an object literal 

```
/Users/jlengrand/IdeaProjects/mergify/lib/utils/logger/index.js:4
  log = (...args) =>{
  ^^^^^^^^^^^^^^^^^^^

SyntaxError: Invalid shorthand property initializer
```

Seems like variable arguments is not supported in the same way in there.

Comment (jlengrand) [reply lib/utils/logger/index.js]: Nevermind, got it working!

**Comment by jlengrand** [comment]
Alright, all what's left now is to test.

**Comment by jlengrand** [comment]
Alright, I have something but I'm not happy about the tests. Can you please have a look at it?

File: lib/utils/logger/index.spec.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            log: function(...args) {
+                this.content = [...args];
+            },
+        }
+        logger = getLogger(mockConsole);
+    });
+
+    test('Testing normal log message', () => {
+
+        logger.log('test', 'plop');
```
Comment (RamonGebben) [inline lib/utils/logger/index.spec.js]: Why not use mock functions from Jest?
https://jestjs.io/docs/en/mock-functions 

Then you could `getLogger` with `{ log: jest.fn() }`
```javascript
const mockConsole = {
  log: jest.fn(),
};

logger = getLogger(mockConsole);

logger.log('Pizza')

expect(mockConsole.log.mock.results[0].value).toBe('Pizza')
```

File: lib/utils/logger/index.js [hunk 8]
```diff
@@ -1,3 +1,22 @@
+const chalk = require('chalk');
+
+const getLogger = function(transport){
+  return {
+    log : (...args) =>{
```
Comment (RamonGebben) [inline lib/utils/logger/index.js]: Formatting can be a little prettier:

```
return {
  log: (...args) => {
  },
}
```

File: index.js [hunk 9]
```diff
@@ -52,6 +60,10 @@ const run = async() => {
     await verify();
   }
 
+  options.forEach(({trigger, description, fn}) => {
```
Comment (RamonGebben) [inline index.js]: Can we put whitespace around deconstruction arguments?
```
({ a, b }) => //...
```
